### PR TITLE
Example/multiclass ecoc

### PR DESCRIPTION
**Pull request by Hephaestus12**

### CODE REVIEW

File: examples/meta/src/multiclass/ecoc_discriminant_aed.sg.in [hunk 1]
```diff
@@ -0,0 +1,40 @@
+CSVFile f_feats_train("@SHOGUN_DATA@/classifier_4class_2d_linear_features_train.dat")
```
Comment (karlnapf) [inline examples/meta/src/multiclass/ecoc_discriminant_aed.sg.in]: you added a new file for chaid tree. need to sort that out

**Review by karlnapf — COMMENTED** [review]
LGTM, only the data commit is somehow wrong

**Comment by gf712** [comment]
@karlnapf Does it make sense to keep the legacy test though? Because it tests for all possible ecoc encoder/decoder combinations

**Comment by karlnapf** [comment]
> @karlnapf Does it make sense to keep the legacy test though? Because it tests for all possible ecoc encoder/decoder combinations

I think what makes sense is to add one meta example for each of the combinations covered in the example

**Comment by Hephaestus12** [comment]
> I think what makes sense is to add one meta example for each of the combinations covered in the example

Yes, I plan to do that, should I include all of them in this PR or separate ones?

**Comment by karlnapf** [comment]
You can do this PR. But wait until we are happy with the one you made here as then you dont need to change all files ... :)

**Comment by karlnapf** [comment]
Actually it looks good so you can create the other ones :)

File: examples/meta/src/multiclass/ecoc_discriminant_hd.sg.in [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+MulticlassStrategy strategy = create_multiclass_strategy("ECOCStrategy", encoder=encoder, decoder=decoder)
+#![choose_strategy]
+
+#![create_instance]
+Machine mc_classifier = create_machine("LinearMulticlassMachine", multiclass_strategy=strategy, machine=classifier, labels=labels_train)
+#![create_instance]
+
+#![train_and_apply]
+mc_classifier.train(features_train)
+MulticlassLabels labels_predict = mc_classifier.apply_multiclass(features_test)
```
Comment (karlnapf) [inline examples/meta/src/multiclass/ecoc_discriminant_hd.sg.in]: could you use `apply`

File: examples/meta/src/multiclass/ecoc_discriminant_hd.sg.in [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#![train_and_apply]
+
+#![evaluate_accuracy]
+Evaluation eval = create_evaluation("MulticlassAccuracy")
+real accuracy = eval.evaluate(labels_predict, labels_test)
+#![evaluate_accuracy]
+
+# integration testing variables
+RealVector output = labels_predict.get_labels()
```
Comment (karlnapf) [inline examples/meta/src/multiclass/ecoc_discriminant_hd.sg.in]: could you use `get` ? (also in all other cases)

**Review by karlnapf — COMMENTED** [review]
thanks!
You used some specific API (I made comments). In general, stick to base class methods and use put/get

**Comment by Hephaestus12** [comment]
> thanks!
> You used some specific API (I made comments). In general, stick to base class methods and use put/get

Okay, I'l make those changes, Will I need to update the data PR again, or just this PR?

**Comment by karlnapf** [comment]
You only need to update the data files if the numbers change. You will see that locally. I think they won't change if you do these changes.

**Comment by Hephaestus12** [comment]
Hi, I replaced it with the new API, I'm getting this error:
```
/home/tejsukhatme/shogun/build/examples/meta/cpp/multiclass/ecoc_discriminant_aed.cpp: In function ‘int main(int, char**)’:
/home/tejsukhatme/shogun/build/examples/meta/cpp/multiclass/ecoc_discriminant_aed.cpp:76:44: error: no matching function for call to ‘shogun::DynamicObjectArray::append_element(std::shared_ptr<shogun::SGObject>&, const char [7])’
 sg_storage->append_element(output, "output");
                                            ^
In file included from /home/tejsukhatme/shogun/build/examples/meta/cpp/multiclass/ecoc_discriminant_aed.cpp:10:0:
/home/tejsukhatme/shogun/src/shogun/lib/DynamicObjectArray.h:198:15: note: candidate: template<class T, class T2> bool shogun::DynamicObjectArray::append_element(T, const char*)
   inline bool append_element(T e, const char* name="")
               ^~~~~~~~~~~~~~
```

How do I avoid this? I am not dealing with the Dynamic Array allocation myself in the meta example.
[This](https://gist.github.com/Hephaestus12/2baef226f958087b1540886b48c583ad) is the meta example.

**Comment by karlnapf** [comment]
Made a comment in the gist. Could you check if that helps?
Also could you pls post the .cpp source of the example?

**Comment by Hephaestus12** [comment]
> Actually just checking your example, the error doesnt match that. Could you re-run `make meta_examples` and then `make build_cpp_meta_examples`....maybe something is outdated. Then add the change I posted, and re-run. If the error persists, post the .cpp

Yes, I did that. The error still persists. These are the relevant files:
[Error Message](https://pastebin.pl/view/d236b586)
[.sg.in file](https://gist.github.com/Hephaestus12/2baef226f958087b1540886b48c583ad)
[.cpp file](https://gist.github.com/Hephaestus12/580db3c1929ef0b50c49f84febf986b0)

**Comment by gf712** [comment]
Something isn’t right, you should be getting an error before that. The issue is that you are getting labels as an sgobject whereas it is an SGVector.. but when you get there should be an error if you pass the wrong type.. I’ll have a look into at some point. Have a look at how other meta examples get an SGVector and you’ll be able to fix this :)

**Comment by Hephaestus12** [comment]
> thanks!
> You used some specific API (I made comments). In general, stick to base class methods and use put/get

Is it okay if I use `get_real_vector()` instead of `get()`? That's what other meta examples seem to do.

**Comment by gf712** [comment]
Yup, that’s a templated version of get for SGVector. You can have a look at the cpp file using that and see it still uses get :)

**Comment by Hephaestus12** [comment]
Is this okay to merge?

**Comment by geektoni** [comment]
> Is this okay to merge?

CI does not seem to be happy. Did you check that the generated meta examples work locally?

File: examples/meta/src/multiclass/ecoc_discriminant_aed.sg.in [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+CSVFile f_feats_test("@SHOGUN_DATA@/classifier_4class_2d_linear_features_test.dat")
+CSVFile f_labels_train("@SHOGUN_DATA@/classifier_4class_2d_linear_labels_train.dat")
+CSVFile f_labels_test("@SHOGUN_DATA@/classifier_4class_2d_linear_labels_test.dat")
```
Comment (gf712) [inline examples/meta/src/multiclass/ecoc_discriminant_aed.sg.in]: `CSVFile` has now been removed from the interfaces, you now need to use `File` and `read_csv` instead. Have a look at the other examples

Comment (Hephaestus12) [reply examples/meta/src/multiclass/ecoc_discriminant_aed.sg.in]: Yes okay, I'll do that. Are there any other changes to be made as well?

Comment (gf712) [reply examples/meta/src/multiclass/ecoc_discriminant_aed.sg.in]: You'll have to see!

**Comment by karlnapf** [comment]
> Something isn’t right, you should be getting an error before that. The issue is that you are getting labels as an sgobject whereas it is an SGVector.. but when you get there should be an error if you pass the wrong type.. I’ll have a look into at some point. Have a look at how other meta examples get an SGVector and you’ll be able to fix this :)

This indeed is concerning. Shouldn't have happened. Especially since this would lead to a segfault down the line when accessing the returned object (it is a SGVector after all)

**Comment by karlnapf** [comment]
Looks like CI passed (using the data you had). Thanks, that is a big chunk! :)

**Pull request merged**